code/game/objects/items/devices/gps.dm

```diff
@@ -17,15 +18,98 @@ var/list/GPS_list = list()
 	var/hide_signal = FALSE		// If true, signal is not visible to other GPS devices.
 	var/can_hide_signal = FALSE	// If it can toggle the above var.
 
+	var/mob/holder
+	var/list/tracking_devices
+	var/obj/compass_holder/compass
+
 /obj/item/device/gps/Initialize()
 	. = ..()
+	compass = new(src)
 	GPS_list += src
 	name = "global positioning system ([gps_tag])"
+	update_holder()
 	update_icon()
 
+/obj/item/device/gps/proc/update_holder()
+
+	if(holder && loc != holder)
+		GLOB.moved_event.unregister(holder, src)
+		GLOB.dir_set_event.unregister(holder, src)
+		holder.client?.screen -= compass
+		holder = null
+
+	if(istype(loc, /mob))
+		holder = loc
+		GLOB.moved_event.register(holder, src, .proc/update_compass)
+		GLOB.dir_set_event.register(holder, src, .proc/update_compass)
+
+	if(holder && tracking)
+		if(!(src in SSobj.processing))
+			START_PROCESSING(SSobj, src)
+		if(holder?.client)
+			if(holder.get_active_hand() == src || holder.get_inactive_hand() == src)
+				holder.client.screen |= compass
+			else
+				holder.client.screen -= compass
+	else
+		STOP_PROCESSING(SSobj, src)
+		if(holder?.client)
+			holder.client.screen -= compass
+
+/obj/item/device/gps/pickup()
+	. = ..()
+	update_holder()
+
+/obj/item/device/gps/equipped()
+	. = ..()
+	update_holder()
+
+/obj/item/device/gps/dropped()
+	. = ..()
+	update_holder()
+
+/obj/item/device/gps/process()
+	if(!tracking)
+		return PROCESS_KILL
+	update_holder()
```

Comment:
What circumstances does this try to handle that the others don't? I'm a little wary of the `if(!(src in SSobj.processing))` on account of not knowing how big that list is, on average.

Reply:
Seems to be unnecessary, I assume it was to handle an edge case that I then found a proc to hook for.

Reply:
Ah found it. equipped() doesn't fire on putting something in your pocket.

Reply:
Hm odd, it should be firing but isn't. If I proc call update_holder when it's in my pocket, it properly removes the overlay.

Reply:
Given that the processing list isn't getting checked endlessly now I think it should be fine to leave update_holder in process to catch edge cases like pockets. Can't work out why they're not updating properly in any case.